### ISSUE TITLE
OpenAI error: `Premature close (ERR_STREAM_PREMATURE_CLOSE)`

### ISSUE THREAD

**Issue by davidfant** [langchain-ai/langchainjs#4196]
Sometimes when running OpenAI streaming I get the following error:
```
Error: Premature close
    at IncomingMessage.<anonymous> (/app/node_modules/node-fetch/lib/index.js:1748:18)
    at Object.onceWrapper (node:events:631:28)
    at IncomingMessage.emit (node:events:517:28)
    at emitCloseNT (node:internal/streams/destroy:132:10)
    at process.processTicksAndRejections (node:internal/process/task_queues:81:21) {
 code: 'ERR_STREAM_PREMATURE_CLOSE'
}
```

I don't know how to repro, but it happens consistently when I run big batches of jobs with many prompts. Sometimes it happens just when starting the request, sometimes it happens a while in (e.g. 10+ seconds into the generation).

`ChatOpenAI.onFailedAttempt` _never_ gets called, so somehow `p-retry` inside `ChatOpenAI` isn't retrying requests when this happens. Instead, the agent runnable that invokes `this.model` catches the error.

Any idea how to make this error caught inside ChatOpenAI so that the request can be retried?

This is how I create my Langchain ChatOpenAI runnable:
```ts
import winston from "winston";

const STATUS_NO_RETRY = [
  400, // Bad Request
  401, // Unauthorized
  402, // Payment Required
  403, // Forbidden
  404, // Not Found
  405, // Method Not Allowed
  406, // Not Acceptable
  407, // Proxy Authentication Required
  408, // Request Timeout
  409, // Conflict
];

function createFailedAttemptHandler(logger?: winston.Logger) {
  return (error: any) => {
    logger?.warn('OpenAI failed attempt', { error, stack: (error as Error).stack });

    if (
      error.message.startsWith('Cancel') ||
      error.message.startsWith('AbortError') ||
      error.name === 'AbortError'
    ) {
      logger?.warn('OpenAI failed because of cancel/abort', { error });
      throw error;
    }

    const status = (error as any)?.response?.status ?? (error as any)?.status;
    if (status && STATUS_NO_RETRY.includes(+status)) {
      logger?.error('OpenAI failed because of status', { status });
      throw error;
    }

    if ((error as any)?.error?.code === 'insufficient_quota') {
      logger?.warn('OpenAI failed because of insufficient quota', { error });
      const err = new Error(error?.message);
      err.name = 'InsufficientQuotaError';
      throw err;
    }
  };
}

this.model = new ChatOpenAI({
      temperature: 0,
      modelName: 'gpt-4-1106-preview',
      maxTokens: 1000,
      streaming: true,
      cache: fields.cache,
      configuration: fields.config,
      maxRetries: 10,
      onFailedAttempt: createFailedAttemptHandler(fields.logger),
    })
```

**Comment by jacoblee93** [comment]
Do you need to have streaming enabled? Catching streaming errors is much more nuanced (e.g. what to do with partial responses).

**Comment by davidfant** [comment]
> Do you need to use `.stream()` for this or could use `.batch()` instead? Catching streaming errors is much more nuanced (e.g. what to do with partial responses).

I need to use stream (because this is streamed to the user on the frontend)! I'd be fine with just rerunning the request if I get a partial response. Do you know how I'd catch the streaming errors in the first place? Right now they seem to be falling through the retry logic

**Comment by davidfant** [comment]
@jacoblee93 I think I see what's going on: when using streaming, `completionWithRetry` is used in `ChatOpenAI._streamResponseChunks`, which `await`s (and should retry) the _creation_ of the stream iterable. However, when later iterating over the stream, any errors thrown there won't be retried. Might that be right?
https://github.com/langchain-ai/langchainjs/blob/6e5d02188bd29c7fe51ff1440a097150f88c547d/libs/langchain-openai/src/chat_models.ts#L463

A somewhat related issue: I'm also getting streams that sometimes end unexpectedly while generating JSON. This usually happens when OpenAI API reqs take an unreasonable amount of time (e.g. 4 min vs the regular 20s), so something is prob wrong on OpenAI's side, but Langchain doesn't allow me to catch these problems anyway. That throws errors in `getEstimatedTokenCountFromPrompt` because JSON parsing fails:
https://github.com/langchain-ai/langchainjs/blob/6e5d02188bd29c7fe51ff1440a097150f88c547d/libs/langchain-openai/src/chat_models.ts#L711-L715

Those errors are also just thrown and not retried because it happens outside of the retry logic.

One potential solution:
* Wrap the full streaming logic in the retry logic
* In the `defaultFailedAttemptHandler`, rethrow the errors that are streaming related: https://github.com/langchain-ai/langchainjs/blob/6e5d02188bd29c7fe51ff1440a097150f88c547d/langchain-core/src/utils/async_caller.ts#L18
In this way, ppl still get the same behavior as now, but for use cases like mine that want more control over error handling, it's possible to provide a custom `onFailedAttempt` handler that e.g. decide to retry these errors.

Any reason you wouldn't want to wrap the whole streaming logic in the retry logic?
https://github.com/langchain-ai/langchainjs/blob/6e5d02188bd29c7fe51ff1440a097150f88c547d/libs/langchain-openai/src/chat_models.ts#L526-L559

**Comment by jacoblee93** [comment]
> @jacoblee93 I think I see what's going on: when using streaming, completionWithRetry is used in ChatOpenAI._streamResponseChunks, which awaits (and should retry) the creation of the stream iterable. However, when later iterating over the stream, any errors thrown there won't be retried. Might that be right?

Yes, this is the intended behavior - we basically catch immediately thrown errors (missing auth tokens) for convenience. When we were implementing the original version of runnables, it wasn't 100% clear what default retry behavior should be mid-stream (abstractly, you could stream from some stateful API with consequences for just re-requesting like that), so we don't do anything by default and just throw it.

I think some way to hook a handler in would be reasonable though - for now, would it unblock you to implement it in your own application logic?

**Comment by jacoblee93** [comment]
I think for token counting logic we should absolutely not error the entire call on parsing failures, so yes that's definitely a bug.

**Comment by davidfant** [comment]
Sure, I'll implement the retry logic myself

**Comment by kahliahogg** [comment]
@davidfant curious if you ended up finding a solution to this you could share?

**Comment by EfrosIonelu** [comment]
any update there ?

**Comment by Cornelius000** [comment]
We are facing a similar issue with just the completion api

**Comment by jacoblee93** [comment]
Should have some mechanism here - will think through it a bit more.

**Comment by irina-alymova** [comment]
Could be caused by https://github.com/node-fetch/node-fetch/issues/1767

**Comment by Float07** [comment]
I'm having the same issue, and I'm also using `node-fetch`, which makes me think this is an issue with `node-fetch` and not with OpenAI, like @irina-alymova suggested.

Anyone tried using a more recent Node version and using the native `fetch` instead of the `node-fetch` library? Maybe that would fix the issue?

It seems that the `node-fetch` issue is still open with not solution, unfortunately: https://github.com/node-fetch/node-fetch/issues/1767
